mesmer/stats/_linear_regression.py

```diff
@@ -319,3 +356,92 @@ def _fit_linear_regression_np(predictors, target, weights=None, fit_intercept=Tr
         intercepts = np.zeros_like(coefficients[:, :1])
 
     return np.hstack([intercepts, coefficients])
+
+
+def prep_linear_regression_data(
+    predictors: DataTree,
+    target: DataTree,
+    stacking_dims: list[str],
+    weights: DataTree | None = None,
+    collapse_dim: str = "scenario",
+    stacked_dim: str = "sample",
+) -> tuple[DataTree, xr.Dataset, xr.Dataset] | tuple[DataTree, xr.Dataset]:
+    """
+    prepares data for Linear Regression:
+    1. Broadcasts predictors to target
+    2. Collapses DataTrees into DataSets
+    3. Stacks the DataSets along the stacking dimensions
+
+    Parameters
+    ----------
+    predictors : DataTree
+        A ``DataTree`` of ``xr.Dataset`` objects used as predictors. The ``DataTree``
+        must have subtrees for each predictor each of which has to have at least one
+        leaf, holding a ``xr.Dataset`` representing a scenario. The subtrees of
+        different predictors must be isomorphic (i.e. have the save scenarios). The ``xr.Dataset``
+        must at least contain `dim` and each ``xr.Dataset`` must only hold one data variable.
+    target : DataTree
+        A ``DataTree``holding the targets. Must be isomorphic to the predictor subtrees, i.e.
+        have the same scenarios. Each leaf must hold a ``xr.Dataset`` which must be at least 2D
+        and contain `dim`, but may also contain a dimension for ensemble members.
+    stacking_dims : list[str]
+        Dimension(s) to stack.
+    weights : DataTree, default: None.
+        Individual weights for each sample, must be isomorphic to target. Must at least contain
+        `dim`, and must have the ensemble member dimesnion if target has it.
+    collapse_dim : str, default: "scenario"
+        Dimension along which to collapse the DataTrees, will automatically be added to the
+        stacking dims.
+    stacked_dim : str, default: "sample"
+        Name of the stacked dimension.
+
+    Returns
+    -------
+    tuple
+        Tuple of the prepared predictors, target and weights, where the predictors and target are
+        stacked along the stacking dimensions and the weights are stacked along the stacking dimensions
+        and the ensemble member dimension.
+
+    Notes
+    -----
+    Dimensions which exist along the target but are not in the stacking_dims will be excluded from the
+    broadcasting of the predictors.
+    """
+
+    stacking_dims.append(collapse_dim)
+
+    # exclude target dimensions from broadcasting which are not in the stacking_dims
+    exclude_dim = set(target.leaves[0].ds.dims) - set(stacking_dims)
+
+    # predictors need to be
+    for key, subtree in predictors.items():
+        # 1) broadcast to target
+        broadcasted = subtree.broadcast_like(target, exclude=exclude_dim)
+        # 2) collapsed into DataSets
+        ds = collapse_datatree_into_dataset(broadcasted, dim=collapse_dim)
```

Comment:
```suggestion
        broadcast = subtree.broadcast_like(target, exclude=exclude_dim)
        # 2) collapsed into DataSets
        ds = collapse_datatree_into_dataset(broadcast, dim=collapse_dim)
```

I think `broadcasted` is not correct

Reply:
😅